cmd/pebble/cmd_logs.go

```diff
@@ -0,0 +1,146 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strconv"
+	"time"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/canonical/pebble/client"
+)
+
+const (
+	logTimeFormat = "2006-01-02T15:04:05.000Z07:00"
+)
+
+type cmdLogs struct {
+	clientMixin
+	Follow     bool   `short:"f" long:"follow"`
+	Output     string `short:"o" long:"output"`
+	NumLogs    string `short:"n" long:"number"`
+	Positional struct {
+		Services []string `positional-arg-name:"<service>"`
+	} `positional-args:"yes"`
+}
+
+var logsDescs = map[string]string{
+	"follow": "Follow (tail) logs for given services until Ctrl-C pressed.",
+	"output": "Output format: \"text\" (default), \"json\" (JSON lines), or \n\"raw\" (copy raw log bytes to stdout and stderr).",
+	"number": "Number of logs to show (before following); defaults to 10.\nIf 'all', show all buffered logs.",
+}
+
+var shortLogsHelp = "Fetch service logs"
+var longLogsHelp = `
+The logs command fetches buffered logs from the given services (or all services
+if none are specified) and displays them in chronological order.
+`
+
+func (cmd *cmdLogs) Execute(args []string) error {
+	var numLogs int
+	switch cmd.NumLogs {
+	case "":
+		numLogs = 10
+	case "all":
+		numLogs = -1
+	default:
+		var err error
+		numLogs, err = strconv.Atoi(cmd.NumLogs)
+		if err != nil || numLogs < 0 {
+			return fmt.Errorf(`expected n to be a non-negative integer or "all", not %q`, cmd.NumLogs)
+		}
+	}
+
+	var writeLog client.WriteLogFunc
+	switch cmd.Output {
+	case "", "text":
+		writeLog = writeLogText
+	case "json":
+		writeLog = writeLogJSON
+	case "raw":
+		writeLog = writeLogRaw
+	default:
+		return fmt.Errorf(`invalid output format (expected "json", "text", or "raw", not %q)`, cmd.Output)
+	}
+
+	opts := client.LogsOptions{
+		WriteLog: writeLog,
+		Services: cmd.Positional.Services,
+		NumLogs:  &numLogs,
+	}
+	var err error
+	if cmd.Follow {
+		// Stop following when Ctrl-C pressed (SIGINT).
+		ctx := notifyContext(context.Background(), os.Interrupt)
+		err = cmd.client.FollowLogs(ctx, &opts)
+	} else {
+		err = cmd.client.Logs(&opts)
+	}
+	return err
+}
+
+// Needed because signal.NotifyContext is Go 1.16+
+func notifyContext(parent context.Context, signals ...os.Signal) context.Context {
+	ctx, cancel := context.WithCancel(parent)
+	ch := make(chan os.Signal)
+	signal.Notify(ch, signals...)
+	go func() {
+		// Wait for signal, then cancel the context.
+		<-ch
+		cancel()
+	}()
+	return ctx
+}
+
+func writeLogText(timestamp time.Time, service, message string) error {
+	suffix := ""
+	if len(message) == 0 || message[len(message)-1] != '\n' {
+		suffix = "\n"
+	}
+	_, err := fmt.Fprintf(Stdout, "%s [%s] %s%s",
+		timestamp.Format(logTimeFormat), service, message, suffix)
+	return err
+}
+
+func writeLogRaw(timestamp time.Time, service, message string) error {
+	_, err := io.WriteString(Stdout, message)
```

Comment:
The idea here is to show the raw bytes written by the service, without the timestamp or service name prefix. This could be useful if you're running a service which already log with timestamps, so you don't want to show the double-timestamp and service name. For example, if you have a service named `snappass` that wrote the following log lines, by default it'd show this:

```
2021-06-08T15:27:34.123Z snappass June 8, 2021 3:27:34PM did something`
```

But with `--format=raw` it'd show this:

```
June 8, 2021 3:27:34PM did something`
```

That said, I'm happy to not include the "raw" option for now if you still don't like it.

Reply:
That sounds like an ugly hack to solve a nasty behavior we have on the system at the moment. I'd prefer to keep it out, and solve the underlying issue of having double timestamps like that. We should do that next, before people start stumbling upon that behavior and thinking we're insane.

Reply:
My usecase for raw was around JSON and yaml dumps, panics/stack traces etc. Sure you could use `cut` to remove those lines so its not the worst, but ideally you wouldn't have to.

Some applications have a json log format that is newline delimited, I'd like to be able to easily support consuming that.

Reply:
Done (also added a TODO item to address this).